larpandoracontent/LArReclustering/ThreeDMultiReclusteringAlgorithm.cc

```diff
@@ -0,0 +1,355 @@
+    {
+        CaloHitList caloHits;
+        pCluster->GetOrderedCaloHitList().FillCaloHitList(caloHits);
+        freeCaloHits.insert(freeCaloHits.end(), caloHits.begin(), caloHits.end());
+        freeCaloHits.insert(freeCaloHits.end(), pCluster->GetIsolatedCaloHitList().begin(), pCluster->GetIsolatedCaloHitList().end());
+        PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::Delete(*this, pCluster, clusterListName));
+    }
+
+    return STATUS_CODE_SUCCESS;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+StatusCode ThreeDMultiReclusteringAlgorithm::BuildNewPfos(const ClusterList &clusters,
+                                                          std::map<HitType, CaloHitList> &viewToFreeCaloHits2D) const
+{
+    // Cluster any 2D hits directly associated with 3D hits in the same way as the 3D clusters
+    std::map<HitType, ClusterList> viewToNewClusters2D = { {TPC_VIEW_U, {}}, {TPC_VIEW_V, {}}, {TPC_VIEW_W, {}} };
+    std::vector<ClusterList> newPfoClusters;
+    for (const Cluster *const pCluster3D : clusters)
+    {
+        ClusterList newClusters;
+        PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, Build2DClustersFrom3D(pCluster3D, viewToFreeCaloHits2D, newClusters));
+        for (const Cluster *const pCluster : newClusters)
+            viewToNewClusters2D.at(LArClusterHelper::GetClusterHitType(pCluster)).emplace_back(pCluster);
+        newClusters.emplace_back(pCluster3D);
+        newPfoClusters.emplace_back(newClusters);
+    }
+
+    // Put any remaining 2D hits into the nearest new 2D cluster made in the last step
+    // If no 2D clusters for a view could be made in the previous step (happens rarely when no 3D hits get made from one of the cluster views),
+    // just mop up the remaining 2D hits into the old 2D clusters
+    // NOTE Hits added to 2D clusters this way are added as isolated
+    for (const auto &[view, caloHits2D] : viewToFreeCaloHits2D)
+    {
+        if (viewToNewClusters2D.at(view).empty())
+        {
+            std::string clusterListName {m_clusterListNames.at(view)};
+            const ClusterList *pOldClusters2D {nullptr};
+            PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::GetList(*this, clusterListName, pOldClusters2D));
+            // This means no hits in this view made any 3D hits, this happens very rarely for events with few hits.
+            // In this case, just leave these hits unclustered. Could also consider putting them in their own isolated hit-only cluster?
+            if (pOldClusters2D->empty())
+                continue;
+            PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, MopUpCaloHits(caloHits2D, *pOldClusters2D, true));
+            continue;
+        }
+        PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, MopUpCaloHits(caloHits2D, viewToNewClusters2D.at(view), true));
```

Comment:
I'd like to understand the motivation for the 'framework' being in charge of mopping up hit that were failed to be reclustered.
To me this feels like a physics/end user decision and not something the framework should decide.

Reply:
Personally I don't think the framework should do the mopping up and it should be left to a future alg.
But, I think this is a bit of a technical detail that shouldn't hold up the PR.

I think the halfway house solution is to add a flag to the host class that enables/disables the mop up as you have it now.  Have the XML read in the value of the flag and set it to true for now.